src/cdk/tree/navigator.ts

```diff
@@ -0,0 +1,311 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+import {FocusableOption} from '@angular/cdk/a11y';
+import {Directionality} from '@angular/cdk/bidi';
+import {
+  DOWN_ARROW,
+  END,
+  ENTER,
+  HOME,
+  LEFT_ARROW,
+  NUMPAD_MULTIPLY,
+  RIGHT_ARROW,
+  TAB,
+  UP_ARROW
+} from '@angular/cdk/keycodes';
+import {Directive, Optional} from '@angular/core';
+import {Subject} from 'rxjs';
+import {TreeControl} from './control/tree-control';
+
+
+/**
+ * Keyboard navigator interface.
+ */
+// TODO(tinayuangao): Move to a common place
+export interface CdkNavigator<T> {
+  insert(index: number, data: T, node: FocusableOption, parentData?: T);
+  remove(index: number, parentData?: any);
+  move(previousIndex: number, currentIndex: number, parentData?: T);
+  updateFocusedData(newFocusedData: T);
+}
+
+/**
+ * Navigator for CDK tree component. Use keyboard to navigate through the tree nodes.
+ * `HOME` moves the focus to first node in the tree
+ * `END` moves the focus to last node in the tree
+ * `UP_ARROW`, and `DOWN_ARROW` moves the focus to previous/next visible node.
+ * `LEFT_ARROW` moves the focus to parent node, or collapse current focused node.
+ * `RIGHT_ARROW` moves the focus to first child node, or expand the current focused node.
+ */
+@Directive({
+  selector: '[cdkTreeNavigator]',
+  host: {
+    '(keydown)': '_handleKeydown($event)',
+    'class': 'cdk-tree-navigator',
+  }
+})
+export class CdkTreeNavigator<T> implements CdkNavigator<T> {
+  /** The node map map data nodes to CdkTreeNodes */
+  protected nodeMap: Map<T, FocusableOption> = new Map<T, FocusableOption>();
+
+  /** A map from parent node to a list of children. */
+  protected nodeListMap: Map<T | undefined, T[]> = new Map<T | undefined, T[]>();
+
+  /** A map from node data to its parent node data. */
+  protected parentMap: Map<T, T> = new Map<T, T>();
+
+  /** Current focused node data. */
+  focusedData: T;
+
+  /** Tree control is used to expand or collapse nodes. */
+  treeControl: TreeControl<T>;
+
+  /**
+   * Stream that emits any time the TAB key is pressed, so components can react
+   * when focus is shifted off of the list.
+   */
+  tabOut: Subject<void> = new Subject<void>();
```

Comment:
I think we should omit the tab bits; it's something in retrospect I don't think belongs in the list key managers either